### PR TITLE
ARROW-835: [Format][C++][Java] Create a new Duration type

### PR DESCRIPTION
**Pull request by emkornfield**
- Create a new DurationInterval type in format.
- Implement all interval types in C++ and Java including integration test.

Once this is checked in, I think https://issues.apache.org/jira/browse/ARROW-352 can be resolved as a  won't fix?

### CODE REVIEW

**Comment by emkornfield** [comment]
@wesm could you look at the format changes to see if you are ok with them?
@pravindra Can you look at the impact on the java side (I added all the necessary java classes I think) to confirm this is acceptable for the Java?

File: format/Schema.fbs
Comment (emkornfield) [reply format/Schema.fbs]: Sorry missed the e-mail on this.  Will fix in a second.  My intent was not to change the java representation which appears to store Days and Tiem separately.

Comment (emkornfield) [reply format/Schema.fbs]: I pushed a new commit, please take a look.

File: java/vector/src/main/java/org/apache/arrow/vector/IntervalEpochMilliVector.java
Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/IntervalEpochMilliVector.java]: Done.  Tried to reword a little bit to make more sense.

File: format/Schema.fbs
Comment (emkornfield) [reply format/Schema.fbs]: Sorry just addressed this comment more fully.

**Comment by emkornfield** [comment]
Failure was NodeJs

**Review by pravindra — APPROVED** [review]

**Comment by pravindra** [comment]
since this is a format change (backward compatible though), I'll wait for @wesm to approve too before merging.

**Comment by emkornfield** [comment]
@pravindra thanks for the review

File: format/Schema.fbs
Comment (emkornfield) [inline format/Schema.fbs]: Looking over the old e-mail threads it might be that TimeDelta or TimestampDelta is a better name for this type?

**Comment by emkornfield** [comment]
@wesm based on comments on the other JIRAs CLs it seems like it would be desirable to add an implementation for C++, I can add that to this CL but would you mind taking a look to see if you are ok with naming/modeling (I would prefer not to have to refactor just due to naming issues if possible).

**Comment by emkornfield** [comment]
I believe the Python on OS/X error is spurious (it was green on previous build).

**Comment by wesm** [comment]
Will work on reviewing the C++ side of this

File: cpp/src/arrow/array.h
Comment (wesm) [inline cpp/src/arrow/array.h]: FWIW, I'm a bit disappointed that we aren't deprecating this type =/

File: cpp/src/arrow/array/builder_time.h
Comment (wesm) [inline cpp/src/arrow/array/builder_time.h]: Is this redundant with `type_fwd.h`?

File: cpp/src/arrow/builder.cc
Comment (wesm) [inline cpp/src/arrow/builder.cc]: I think `static_cast` (or `checked_cast`) is sufficient here?

File: cpp/src/arrow/type.h
Comment (wesm) [inline cpp/src/arrow/type.h]: nit: as far as naming I could see an argument for using

* `interval_duration`
* `interval_day_time`
* `interval_month`

but I'm OK with the way things are here

File: format/Schema.fbs
Comment (wesm) [inline format/Schema.fbs]: Got it. If it isn't being deprecated then it's reasonable to implement it in C++ and have integration tests

File: format/Schema.fbs
Comment (wesm) [reply format/Schema.fbs]: Are you happy with "DurationInterval"? It could also be called simply "Duration" or "TimeDuration"

**Review by wesm — COMMENTED** [review]
The C++ side of this, and the changes to the Flatbuffers files, all looks good to me. Thanks @emkornfield for slogging through this and implementing C++, Java, and integration tests in one go!

One question I have is whether we want to commit to the name "DurationInterval" rather than alternatives such as "Duration" or "TimeDelta" or "TimeDuration". I don't think the names are sacred

Can someone (@siddharthteotia ?) carve out a little time to review the Java so we can try to get this merged soon?

**Comment by emkornfield** [comment]
@wesm Thanks for the review I'll try to address comments in the next few days.  In terms of naming, I don't know the etiquette for changing the name once it gets approved on the ML.  If was going to change it, I think Duration would be my choice.

@pravindra already reviewed most of the java code, the main changes that happened where plumbing through support for the integration test.

File: java/vector/src/main/codegen/data/ValueVectorTypes.tdd
Comment (emkornfield) [reply java/vector/src/main/codegen/data/ValueVectorTypes.tdd]: Potentially, I followed the pattern already established in the codebase, but I'm open to trying if it is desirable.  @pravindra @siddharthteotia do you think this is desirable/possible?

File: cpp/src/arrow/array.h
Comment (emkornfield) [reply cpp/src/arrow/array.h]: I didn't want to get dragged into what it means deprecate this, should be easy enough to do if we want it.  As noted below I tried to say it was optional which I seem to recall seeing this was ok on a previous ML thread.

File: java/vector/src/main/codegen/data/ValueVectorTypes.tdd
Comment (jacques-n) [reply java/vector/src/main/codegen/data/ValueVectorTypes.tdd]: Sounds okay to me. Let's see what @siddharthteotia and @pravindra say.

File: cpp/src/arrow/array/builder_time.h
Comment (emkornfield) [reply cpp/src/arrow/array/builder_time.h]: Yes, it appears so, I'll push a commit removing them.  It looks like there might be more redundancy in builder_primitive.h

File: cpp/src/arrow/array/builder_dict.cc
Comment (kou) [inline cpp/src/arrow/array/builder_dict.cc]: Just for your information.

These changes will be conflicted with #4255.
If #4255 is merged, we will be able to just remove these changes.

File: java/vector/src/main/codegen/data/ValueVectorTypes.tdd
Comment (pravindra) [reply java/vector/src/main/codegen/data/ValueVectorTypes.tdd]: IMO, having a single class is definitely desirable. Single this one is starting on a fresh slate, it's worth a try.

File: cpp/src/arrow/type.h
Comment (emkornfield) [reply cpp/src/arrow/type.h]: I could see this to, but I'm happier without the reverse notation.

**Comment by emkornfield** [comment]
@wesm I think I addressed all your feedback.  The OS/X build seems to in an infinite CMake loop with something like 

```For compatibility with older versions of CMake, the install_name fields for
  the following targets are still affected by RPATH settings:
   arrow_python_shared
   arrow_shared
   arrow_testing_shared
   parquet_shared
   plasma_shared
```
I'm not sure why my change would have caused this.  Any thoughts?

@pravindra I was able to consolidate all the java vectors into one, could you take a look?

File: cpp/src/arrow/array/builder_dict.cc
Comment (emkornfield) [reply cpp/src/arrow/array/builder_dict.cc]: Thanks, they were removed.

**Comment by wesm** [comment]
@pravindra @emkornfield where do things stand in this patch? Since I have some changes incoming in ipc/json-internal.cc and ipc/metadata-internal.cc it would be good to merge soon so I can address rebase conflicts

**Comment by pravindra** [comment]
the changes lgtm.

**Comment by emkornfield** [comment]
Thanks @pravindra.  @wesm if you are happy with the changes I think it can merged.

**Comment by emkornfield** [comment]
Also if there is a way to force appveyor to run it would make me feel a bit more comfortable as windows has bitten me in the past

**Comment by wesm** [comment]
@emkornfield usually a `git commit --amend && git push -f` will do the trick, do you want to try that?

**Comment by wesm** [comment]
Baffling

**Comment by wesm** [comment]
It must have something to do with this being an older PR. Is Appveyor enabled on your fork? If not I can push this to my work to get an Appveyor build going

**Comment by emkornfield** [comment]
No appveyor isn't enabled on my fork.  let me try squashing everything, see if that helps.

**Comment by emkornfield** [comment]
Ahh, squash seems to have worked, lets hope things stay green.

File: cpp/src/arrow/type.h
Comment (emkornfield) [reply cpp/src/arrow/type.h]: note "duration" conflicts with some of our time classes so I renamed to "duration_type", hopefully that is ok

Comment (wesm) [reply cpp/src/arrow/type.h]: "some of our time classes"?

Comment (emkornfield) [reply cpp/src/arrow/type.h]: Need to look back I think it was in the vendored time library we were using.  I can look deeper if you don't think there should be a conflict

Comment (emkornfield) [reply cpp/src/arrow/type.h]: looking again they shouldn't conflict, I will try to fix this today.

Comment (wesm) [reply cpp/src/arrow/type.h]: Awesome, thanks.

Comment (emkornfield) [reply cpp/src/arrow/type.h]: renamed.  had to touch some of the vendored code.

**Comment by wesm** [comment]
@emkornfield it looks you made some changes to vendored code which I think we should try to avoid doing in general, can we open a JIRA to restore to the vendored version and address the name conflict another way? I don't want this issue to block the patch since I need this to go in and then rebase #4316 on top

**Comment by wesm** [comment]
Appveyor appears to be about ~6 hours backed up so I'm going to push a branch on my fork to try to get a quicker thumbs up to merge this

**Comment by wesm** [comment]
Here's an Appveyor build starting now on my fork which will take a while to run https://ci.appveyor.com/project/wesm/arrow/builds/24577591

**Comment by emkornfield** [comment]
Thanks.  https://issues.apache.org/jira/browse/ARROW-5346 opened to revert the changes.  Right now the only solution seems like reordering includes in some cases or fixing upstream.  I will have to think about this.

**Comment by wesm** [comment]
Appveyor build looks OK, I'm going to wait a little while yet to make sure the MinGW build works before merging

**Comment by emkornfield** [comment]
@wesm appveyor on your branch looks all greeen.

**Comment by wesm** [comment]
Awesome. Thanks @emkornfield!!

**Pull request closed without merging**
